### ISSUE TITLE
handleSubmit swallows errors thrown in onSubmit functions

### ISSUE THREAD

**Issue by dbryand** [erikras/redux-form#1859]
I was having an issue where I was getting SET_SUBMIT_FAILED on every submit, even though my form had no validation errors. The error would come back empty. After digging into `handleSubmit`, it turns out that the `try` here will catch any type of error and will only give you error messages if caught error has `submitError.errors`.

https://github.com/erikras/redux-form/blob/master/src/handleSubmit.js#L20
```javascript
      try {
        result = submit(values, dispatch, props);
      } catch (submitError) {
        var error = submitError instanceof _SubmissionError2.default ? submitError.errors : undefined;
```

In my case, I had a `type` error in my `onSubmit` function that was very hard to track down. Should we just rethrow if the error is a general runtime error? Something like this:

```javascript
      try {
        result = submit(values, dispatch, props);
      } catch (submitError) {
        var error = submitError instanceof _SubmissionError2.default ? submitError.errors : undefined;
        if (!error) throw submitError;
```

**Comment by naw** [comment]
I've been wondering about this too.

I think rethrowing errors that aren't `SubmissionError` objects is a great idea, but perhaps I'm missing some important scenarios where doing that would be problematic.

_If_ the user has defined a `onSubmitFail` function, then perhaps it is better to pass the error to that callback rather than rethrowing the error, but then the `onSubmitFail` function wouldn't know whether it will be receiving a JavaScript error or a hash of validation errors, which would complicate the `onSubmitFail` API.

At any rate, as it stands now, if you have a legitimate non-validation error in your `onSubmit` function, you have no way of seeing the error unless you wrap your `onSubmit` function in a try/catch and log the exception. This is particularly cumbersome during development where errors from typos or undefined references are common.

**Comment by dbryand** [comment]
Actually, passing the error to onSubmitFail was what I originally expected.
I like that idea best so far.

**Comment by dallonf** [comment]
Just ran into this. Would greatly appreciate a solution that surfaces errors thrown in onSubmit instead of failing silently!

**Comment by dbryand** [comment]
After looking at it more, I do think the simplest, most elegant solution is to just set `error` to `submitError` and let it get pushed into the `onSubmitFail` in user space. Happy to do a PR if we like that.

```
      try {
        result = submit(values, dispatch, props);
      } catch (submitError) {
        var error = submitError instanceof _SubmissionError2.default ? submitError.errors : submitError;
```

**Comment by dbryand** [comment]
On second thought, I think an error should be thrown if `result = submit(values, dispatch, props);` immediately. If we weren't using try/catch, this would be what would happen and I don't think we should be propagating the runtime errors from user code into callbacks, I think it should blow up immediately.

**Comment by bertjwregeer** [comment]
Related: https://github.com/erikras/redux-form/issues/1709

**Comment by akhayoon** [comment]
Is there a workaround for this issue? It would be nice to see the errors.

**Comment by poyannabati** [comment]
@akhayoon If you wrap your passed function in a try/catch you can print the errors before they're caught and swallowed.

**Comment by appden** [comment]
Just ran into this. It doesn't just swallow the error, but the call to `stopSubmit(error)` actually sets `submitSucceeded` to `true`. I needed to manually rethrow a `SubmissionError` with something inside the `errors` property to avoid this.

**Comment by bertjwregeer** [comment]
Interestingly enough I was looking at this bug for another reason 4 days ago (can't remember exactly why), but ran into the issue of not knowing why my form wasn't showing SubmissionError's.

Apparently I had forgotten to import SubmissionError, and since it was silently swallowing errors raised from my Promise I didn't realise that was going on... Trying to code tired is probably also not a good idea, but that threw me for a loop!

**Comment by pierreio** [comment]
@akhayoon It's possible, but error prone, and then you have to handle SubmissionError separately.

**Comment by Urthen** [comment]
Might be going against the grain here, but I'd honestly expect anything outside `SubmissionError` (or whatever other validation errors) to be re-thrown, rather than caught and handed to `onSubmitFail`. At the very least, it should probably be passed to a separate function (like `onSubmitError`)

Just ran into this today, took quite a bit of debugging before realizing handleSubmit was obscuring all errors.

**Comment by jmn8718** [comment]
I have the same issue, but even when i get to send a submission object, the validation `const error = submitError instanceof SubmissionError ? submitError.errors : undefined` doesnt work because the compiled file has this code `submitError instanceof _SubmissionError2.default`.
I know they are the same class, it is just the webpack build, but the typeof my submissionError it is different from an object created with _SubmissionError2.default, and the are both 'SubmissionError'

**Comment by appden** [comment]
I ran into another issue with `stopSubmit()` setting `submitSucceeded` to `true`, which is that I do a transition to another page when `submitSucceeded == true`, which unmounts the form (thus destroying it). Then the line after `stopSubmit()` is `setSubmitSucceeded()`, which will set `submitSucceeded` on the destroyed form. This means when going **back** to the previous page, `submitSucceeded == true` and it will attempt the transition _again_. Hope that made sense. :smile:

**Comment by michabu** [comment]
I also ran into this issue. Took me (with little javascript debugging experiencs) about 2 hours to figure out that `handleSubmit` was silently swalloing the error. Could have been a matter of 2mins otherwise. I always advocate rethrowing / not cathing exceptions that you cannot handle. So I would be really grateful if this could be fixed. I think it would be rather strange to have to register an ExceptionListener in order to get to the Exception. That could be a valid scenario for production code, though.

**Comment by lambdacomplete** [comment]
Ran into the same problem. The exception was thrown in a reducer's case and I had to first figure out whether the case was reached, then at which point it "stopped", figure out why and eventually figure out why the exception was being swallowed and by who. This was tricky since nothing on the UI was changing and nothing was logged in the console.

**Comment by nmaves** [comment]
Agree with most in this issue.  We need to re-throw the exceptions.

**Comment by erikras** [comment]
Fix published in [`v6.2.0`](https://github.com/erikras/redux-form/releases/tag/v6.2.0).
